### PR TITLE
Introduce isolated units for input spec

### PR DESCRIPTION
**Pull request by jeniawhite**
## What does this PR do?
Introduce a new flag (isolate units) in the spec that will allow to disable grouping of same input type inputs into a single component.
If that flag is being activated the inputs won't be grouped and we will create a separate component for each input that will run units for that particular input.

## Why is it important?
This will allow to run multiple inputs of the same type as separate processes/components that handle a single input instead of having a single process/component with multiple inputs represented as units as of today.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## Related issues
- Closes https://github.com/elastic/security-team/issues/8669

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: pkg/component/input_spec.go [hunk 1]
```diff
@@ -21,8 +21,9 @@ type InputSpec struct {
 	Shippers       []string    `config:"shippers,omitempty" yaml:"shippers,omitempty"`
 	Runtime        RuntimeSpec `config:"runtime,omitempty" yaml:"runtime,omitempty"`
 
-	Command *CommandSpec `config:"command,omitempty" yaml:"command,omitempty"`
-	Service *ServiceSpec `config:"service,omitempty" yaml:"service,omitempty"`
+	Command   *CommandSpec `config:"command,omitempty" yaml:"command,omitempty"`
+	Service   *ServiceSpec `config:"service,omitempty" yaml:"service,omitempty"`
+	SkipGroup bool         `config:"skip_group,omitempty" yaml:"skip_group,omitempty"`
```
Comment (blakerouse) [inline pkg/component/input_spec.go:26]: Not a fan of the naming of `skip_group`. How about `isolate_units: true`?

File: pkg/component/component.go [hunk 2]
```diff
@@ -411,22 +428,105 @@ func (r *RuntimeSpecs) componentForInputType(
...
+			InputType:  inputType,
+			OutputType: output.outputType,
+			Units:      units,
+			Features:   featureFlags.AsProto(),
+			Component:  componentConfig.AsProto(),
+			ShipperRef: shipperRef,
+		})
+	}
+
+	return components
```
Comment (blakerouse) [inline pkg/component/component.go:509]: This duplicate lots of code, possible to make this adjustment smaller in the same code above without duplicating all this logic?

**Comment by jeniawhite** [comment]
During the development, I've noticed that there were issues with the `fake` tests.

Specifically with the following conditions:
- Configuring multiple (2) inputs of the same type (isolated units turned on in spec)
- Configuring a shipper for the output
- Running an agent on Windows

This creates a scenario when the agent will spawn a separate component for each input (2), each component will have two units (one output unit and one input unit):
```
{
					ID:         "cloudbeat-default-cloudbeat-0",
					InputType:  "cloudbeat",
					OutputType: "elasticsearch",
					InputSpec: &InputRuntimeSpec{
						InputType:  "cloudbeat",
						BinaryName: "cloudbeat",
						BinaryPath: filepath.Join("..", "..", "specs", "cloudbeat"),
					},
					Units: []Unit{
						{
							ID:       "cloudbeat-default-cloudbeat-0",
							Type:     client.UnitTypeOutput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"type": "shipper",
							}),
						},
						{
							ID:       "cloudbeat-default-cloudbeat-0-unit",
							Type:     client.UnitTypeInput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"type": "cloudbeat",
								"id":   "cloudbeat-0",
							}),
						},
					},
					ShipperRef: &ShipperReference{
						ShipperType: "shipper",
						ComponentID: "shipper-default",
						UnitID:      "cloudbeat-default-cloudbeat-0",
					},
				},
				{
					ID:         "cloudbeat-default-cloudbeat-1",
					InputType:  "cloudbeat",
					OutputType: "elasticsearch",
					InputSpec: &InputRuntimeSpec{
						InputType:  "cloudbeat",
						BinaryName: "cloudbeat",
						BinaryPath: filepath.Join("..", "..", "specs", "cloudbeat"),
					},
					Units: []Unit{
						{
							ID:       "cloudbeat-default-cloudbeat-1",
							Type:     client.UnitTypeOutput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"type": "shipper",
							}),
						},
						{
							ID:       "cloudbeat-default-cloudbeat-1-unit",
							Type:     client.UnitTypeInput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"type": "cloudbeat",
								"id":   "cloudbeat-1",
							}),
						},
					},
					ShipperRef: &ShipperReference{
						ShipperType: "shipper",
						ComponentID: "shipper-default",
						UnitID:      "cloudbeat-default-cloudbeat-1",
					},
				},
				{
					ID:         "shipper-default",
					OutputType: "elasticsearch",
					ShipperSpec: &ShipperRuntimeSpec{
						ShipperType: "shipper",
						BinaryName:  "shipper",
						BinaryPath:  filepath.Join("..", "..", "specs", "shipper"),
					},
					Units: []Unit{
						{
							ID:       "cloudbeat-default-cloudbeat-0",
							Type:     client.UnitTypeInput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"id":   "cloudbeat-default-cloudbeat-0",
								"type": "shipper",
								"units": []interface{}{
									map[string]interface{}{
										"id": "cloudbeat-default-cloudbeat-0-unit",
										"config": map[string]interface{}{
											"type": "cloudbeat",
											"id":   "cloudbeat-0",
										},
									},
								},
							}),
						},
						{
							ID:       "cloudbeat-default-cloudbeat-1",
							Type:     client.UnitTypeInput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"id":   "cloudbeat-default-cloudbeat-1",
								"type": "shipper",
								"units": []interface{}{
									map[string]interface{}{
										"id": "cloudbeat-default-cloudbeat-1-unit",
										"config": map[string]interface{}{
											"type": "cloudbeat",
											"id":   "cloudbeat-1",
										},
									},
								},
							}),
						},
						{
							ID:       "shipper-default",
							Type:     client.UnitTypeOutput,
							LogLevel: defaultUnitLogLevel,
							Config: MustExpectedConfig(map[string]interface{}{
								"type": "elasticsearch",
							}),
						},
					},
				},
```

There is also a component for the shipper (`shipper-default`).
The shipper component has two input units for each input that was configured (`cloudbeat-default-cloudbeat-0`, `cloudbeat-default-cloudbeat-1`).
As far as I understand shipper's server address is based on the component id (encoding, hashing, slicing it) of the shipper.
For each input of the shipper component we are attempting to create a listener to the provided server address.
When running the E2E fake tests, I've noticed that I'm getting an error during the creation of the listeners (with the preconditions mentioned above).
In the following log you will be able to see the creation of the listeners and the failure on listening to the `npipe` (access denied error):
```
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:01.944Z","message":"starting GRPC fake shipper server","component":{"binary":"fake-shipper","dataset":"elastic_agent.fake_shipper","id":"fake-shipper-default","type":"fake-shipper"},"log":{"source":"fake-shipper-default"},"level":"info","server":"npipe:///kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock","time":"2024-04-08T07:48:01Z","ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:01.954Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":624},"message":"Unit state changed fake-default (STARTING->HEALTHY): Healthy","log":{"source":"elastic-agent"},"component":{"id":"fake-shipper-default","state":"HEALTHY"},"unit":{"id":"fake-default","type":"input","state":"HEALTHY","old_state":"STARTING"},"ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:01.954Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":624},"message":"Unit state changed fake-shipper-default (STARTING->HEALTHY): Healthy","log":{"source":"elastic-agent"},"component":{"id":"fake-shipper-default","state":"HEALTHY"},"unit":{"id":"fake-shipper-default","type":"output","state":"HEALTHY","old_state":"STARTING"},"ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:02.251Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":616},"message":"Spawned new unit fake-non-grouped-default-fake-non-grouped: Starting","log":{"source":"elastic-agent"},"component":{"id":"fake-shipper-default","state":"HEALTHY"},"unit":{"id":"fake-non-grouped-default-fake-non-grouped","type":"input","state":"STARTING"},"ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:02.253Z","message":"unit change received","component":{"binary":"fake","dataset":"elastic_agent.fake","id":"fake-default","type":"fake"},"log":{"source":"fake-default"},"expectedState":"HEALTHY","time":"2024-04-08T07:48:02Z","level":"info","state":"CONFIGURING","ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:02.253Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":624},"message":"Unit state changed fake-default-fake (CONFIGURING->HEALTHY): Healthy","log":{"source":"elastic-agent"},"component":{"id":"fake-default","state":"HEALTHY"},"unit":{"id":"fake-default-fake","type":"input","state":"HEALTHY","old_state":"CONFIGURING"},"ecs.version":"1.6.0"}
    log.go:96: {"log.level":"info","@timestamp":"2024-04-08T07:48:02.259Z","message":"starting GRPC fake shipper server","component":{"binary":"fake-shipper","dataset":"elastic_agent.fake_shipper","id":"fake-shipper-default","type":"fake-shipper"},"log":{"source":"fake-shipper-default"},"level":"info","server":"npipe:///kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock","time":"2024-04-08T07:48:02Z","ecs.version":"1.6.0"}
    log.go:96: {"log.level":"error","@timestamp":"2024-04-08T07:48:02.259Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":624},"message":"Unit state changed fake-non-grouped-default-fake-non-grouped (STARTING->FAILED): Error: failed to listen on the named pipe \\\\.\\pipe\\kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock: open \\\\.\\pipe\\kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock: Access is denied.","log":{"source":"elastic-agent"},"component":{"id":"fake-shipper-default","state":"HEALTHY"},"unit":{"id":"fake-non-grouped-default-fake-non-grouped","type":"input","state":"FAILED","old_state":"STARTING"},"ecs.version":"1.6.0"}
    fake_test.go:162: 
        	Error Trace:	C:/Users/windows/agent/testing/integration/fake_test.go:162
        	Error:      	Received unexpected error:
        	            	elastic-agent logged an unexpected error: Unit state changed fake-non-grouped-default-fake-non-grouped (STARTING->FAILED): Error: failed to listen on the named pipe \\.\pipe\kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock: open \\.\pipe\kesXHYHYbU863cFUmh6U9WgsBwPZQPzL.sock: Access is denied.
        	Test:       	TestFakeComponent
--- FAIL: TestFakeComponent (27.47s)
```
This happens only when running the cases on **windows** and also when I've attempted to run the pre-existing `fake` test cases with two different types of inputs having a single input configured the behavior still the same.

When running the same cases on **Darwin** and **Linux** everything works fine and all of the test cases are passing.
This is due to the fact that in such cases we are creating **unix** sockets and they behave differently to **npipe** (named pipes).

I suspect that this isn't related to my changes and running two different components with an input unit each and and output unit for the shipper that use the same shipper on **Windows**, will create the same **npipe** behaviors.

Sadly, I currently do not have an access to a **Windows** machine in order to verify that. This is why I've replicated that with the testing infrastructure using `fakes` (but still one of the `fakes` is the isolated units that deploys a single component with a single input and output unit).

`Cloudbeat` isn't deployed on **Windows** and it is currently the only spec that will use the isolated units option.
It also doesn't use **shippers** and I've added this to it's spec in order to get more coverage from the tests and make sure to **reach 100% of coverage for the code**.

I wouldn't want to block this PR due to that and not sure if that should be resolved as part of it.
Would like your input on that.

Thank you.

File: pkg/component/input_spec.go (hunk 1 above)
Comment (jeniawhite) [reply pkg/component/input_spec.go]: Renamed and changed the terminology across the code.

File: pkg/component/component.go (hunk 2 above)
Comment (jeniawhite) [reply pkg/component/component.go:479]: Refactored code (100% coverage and 0% duplication on new code).

**Comment by blakerouse** [comment]
@jeniawhite I believe this issue is only an issue with the fake shipper code and that is what is affecting your test. The shipper should get 2 input units, but it should only start one GRPC server for those 2 units. At the moment it is trying to start 2 GRPC units at the same address. This is clearly not a problem on Linux and macOS, I don't know why its not. I would have expected it to also cause some connection issues. On Windows its clear that it prevents it from starting another GRPC server at the same npipe address.

Overall I don't believe this is an issue your code or even how the Elastic Agent computes the model. This is only an issue in the fake shipper module.

The fix is to adjust the fake shipper to take both units and then start a single GRPC server. That will fix the issue you are having.

**Comment by jeniawhite** [comment]
@blakerouse 
Added a "caching" mechanism for the servers and reference counters for the units that are using the server.
The access to the cache should be concurrency safe due to the `unitsMx` mutex, but I can also implement the map as a sync map if needed just to be extra safe.

@pchila 
I can see that the CI fails during the run of `TestComponentBuildHashInDiagnostics`.

Due to:
```
testing.go:1225: TempDir RemoveAll cleanup: remove C:\Users\windows\AppData\Local\Temp\TestComponentBuildHashInDiagnostics1045698008\003\state.yaml: The process cannot access the file because it is being used by another process.
```

I assume that these PRs should address that problem:
- https://github.com/elastic/elastic-agent/pull/4596
- https://github.com/elastic/elastic-agent/pull/4598

Please let me know if I'm missing anything.
Thank you.

**Comment by pchila** [comment]
> @pchila I can see that the CI fails during the run of `TestComponentBuildHashInDiagnostics`.
> 
> Due to:
> 
> ```
> testing.go:1225: TempDir RemoveAll cleanup: remove C:\Users\windows\AppData\Local\Temp\TestComponentBuildHashInDiagnostics1045698008\003\state.yaml: The process cannot access the file because it is being used by another process.
> ```
> 
> I assume that these PRs should address that problem:
> 
>     * [close state file in TestComponentBuildHashInDiagnostics #4596](https://github.com/elastic/elastic-agent/pull/4596)
> 
>     * [[8.14](backport #4596) close state file in TestComponentBuildHashInDiagnostics #4598](https://github.com/elastic/elastic-agent/pull/4598)
> 
> 
> Please let me know if I'm missing anything. Thank you.

@jeniawhite correct. Now both PRs are merged, so if you rebase/merge the integration test should be green

**Pull request merged**